File: leetcode/713-subarray-product-less-than-k.py
# https://leetcode.com/problems/subarray-product-less-than-k/

from typing import List

class Solution:
    def numSubarrayProductLessThanK(self, nums: List[int], k: int) -> int:
        result = 0
        window = 1
        j = 0
        for i in range(len(nums)):
            window *= nums[i]
            while window >= k and j <= i:
                window /= nums[j]
                j += 1
            result += (i - j + 1)
        
        return result

File: leetcode/713-subarray-product-less-than-k.js
// https://leetcode.com/problems/subarray-product-less-than-k/

/**
 * @param {number[]} nums
 * @param {number} k
 * @return {number}
 */
var numSubarrayProductLessThanK = function(nums, k) {
  let result = 0;
  let window = 1;
  let j = 0;
  
  for (let i = 0; i < nums.length; i += 1) {
    window *= nums[i];
    while (j <= i && window >= k) window /= nums[j++];
    result += (i - j + 1);
  }

  return result;
};
